### PR TITLE
Adds zreducer

### PR DESCRIPTION
**Pull request by marimeireles**
This PR is going to have a lot of experimental commits. Mainly to keep Johan on the loop =)

### CODE REVIEW

**Review by marimeireles — COMMENTED** [review]
Hey @JohanMabille here are some questions about the dispatcher stuff.
Also is this the way to go to implement the macros?
Thanks!

File: include/zarray/zreducer.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        CT m_e;
+        mutable shape_type m_cache_shape;
+        mutable bool m_cache_initialized;
+        //i have to create some kind of m_args
+        //        Func m_func;
+        //AX m_axes;
+        //OX m_options;
+        //are all of these args gonna be inside of my m_args?
+        detail::strided_view_args<detail::no_adj_strides_policy> m_args;
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/zarray/zreducer.hpp:62]: The elements that should be dispatched are those wrapped in a zarray. For the reducer (like the view), this means the underlying expression only.

File: include/zarray/zreducer.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    };
+
+    /************************
+     * zreducer implementation *
+     ************************/
+    
+    template <class CT, class Func, class AX, class OX>
+    template <class FuncA, class CTA, class AXA, class OXA>
+    inline zreducer<CT>::zreducer(FuncA&& func, CTA&& e, AXA&& axes, OXA&& options)noexcept
+    //do i need forward here
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/zarray/zreducer.hpp:72]: Yes, this is required for `e`. For other arguments, it's a convenient optimization.

File: include/zarray/zreducer.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        : m_e(std::forward<CTA>(e))
+        : m_func(std::forward<FuncA>(func))
+        : m_axes(std::forward<AXA>(axes))
+        : m_options(std::forward<OXA>(options))
+    {
+        auto shape = e.shape();
+        auto offset = e.get_implementation().get_offset();
+        auto layout = e.layout();
+        m_args.fill_args(shape, strides, offset, layout, slices);
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/zarray/zreducer.hpp:81]: axes and options. You should dispatch on `e`  only, `func` should probably be handled th same way we handle functors for operations (a zfunctor mapped to an xfunctor and used an template argument of the dispatcher class)

File: include/zarray/zreducer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (marimeireles) [reply include/zarray/zreducer.hpp:62]: @JohanMabille 
ok, so the `xreducer` on xtensor doesn't have a type `strided_view_args` for itself.
Meaning I'll have to create something like [this](https://github.com/xtensor-stack/xtensor/blob/be8fe0891b869946540505b39ba914096f3f0ceb/include/xtensor/xstrided_view_base.hpp#L820), right? It seems kinda complex, is this really the way to go? and... I'm not really sure where I should make these changes, maybe on the `xreducer.hpp` file that I'm working on?
Thanks!

[the remark this answers is not in the document]
Comment (marimeireles) [reply include/zarray/zreducer.hpp:62]: I'll start with a simple struct

**Comment by JohanMabille** [comment]
Hi @marimeireles sorry for the late reply. It appeared that using the dispatchers for building views required evaluating the underlying expression, which might not play well with chunked arrays and big data. Thus we totally changed the approach and decided to have the wrapper build the view. I'm afraid we need to do the same here, that is, forget the dispatchers, and implement a `buld_reducer` method in the different `zarray_impl` classes, similarly to what has been done for the in #28 .

**Pull request closed without merging**